Test/Microsoft.ApplicationInsights.Test/Shared/TelemetryClientTest.cs

```diff
@@ -140,6 +140,7 @@ public void TrackMetricSendsSpecifiedAggregatedMetricTelemetry()
             var sentTelemetry = new List<ITelemetry>();
             var client = this.InitializeTelemetryClient(sentTelemetry);
 
+#pragma warning disable CS0618 // Type or member is obsolete
```

Comment:
The logic behind deprecating TrackMetric(..):
The recommended new API for metrics is now .GetMetric(..).TrackValue(..). The old TrackMetric(..) API is discouraged because it
a) Distracts from discovering the new API. (Being used to the TrackXxx(..) pattern users will discover it in intellisense and stop looking for other ways to send a metric.)
b) Promotes incorrect coding behavior. (Being used to the TrackXxx pattern users will just construct MetricTelemetry items for each data point and utilize the old API without ever reasoning about the need for pre-aggregation.)

The old TrackMetric(..) API can be used in 2 cases:
A) "Einstein" users who have implemented their own pre-aggregation. These few very advanced users can be expected to realize that TrackMetric(MetricTelemetry item) is just a thin wrapper around Track(ITelemetty item) and that the latter can be used to send the results of their custom pre-aggregation.
B) Users who want to send a single (non-pre-aggregated) measurement. However, such data is not a correct use of AppInsights' data model. If the customer really requires to send and store a single numerical observation, a custom event is the correct data item to use. If the customer is actually interested in a metric, than pre-aggregation is more appropriate. In fact, metrics (on contrary to events) should *never* leave the collection box without pre-aggregation.

Reply:
Discouraging doesn't mean you need to deprecate it. You can make it non editor browseable. I was OK obsoleting it when we discussed introducing new TrackAggregaterMetric API. Since we still have a VALID scenario when one will want to use API - we should not mark it as obsolete.

Reply:
Typical usages of TrackMetric are the "wrong" ones. It is hardly likely that most users wrote their own pre-aggregators and much more likely that they are sending non-preaggregated values. Deprecating the method will highlight the problem to those customers and tell them how to fix it. it will also prevent them from making this mistake in the future. The few Einstein Pre-Aggregation authors are advanced users who should not be too inconvenienced by switching to Track. Especially, because the fact that they fall into this category indicates that they will understand why we are doing this.

Reply:
I have not heard back in 3 days on this particular issue, so I am assuming that my explanation makes sense.
Notably, a beta release is the right place to deprecate this API and to see what customer feedback we get from it.

Reply:
As a non-maintainer of this SDK who doesn't have a ton of context for why this API is the way it is, I'm curious at why these changes are incompatible with `TrackMetric`

@macrogreg You noted `TrackMetric` should be marked obsolete because it:
> Promotes incorrect coding behavior. (Being used to the TrackXxx pattern users will just construct MetricTelemetry items for each data point and utilize the old API without ever reasoning about the need for pre-aggregation.)

Isn't this entirely our fault? `TrackMetric` does whatever we say it does... If it should have been aggregating the whole time why don't we fix the disconnect from user expectations and actual behavior by making `TrackMetric` work the _right_ way? Why is it important that users of our `TelemetryClient` API need to reason separately over metrics/aggregation and use different API methods that are inconsistent from standard `TelemetryClient` methods for tracking these when we can transparently aggregate it behind the scenes?

Is the reason for all of this so users can specify the type of metric / means of aggregation? Can't that be an optional parameter to `TrackMetric` and it can handle all the necessary plumbing with the aggregator internally?

Reply:
Hi @OsvaldoRosado ,

As this topic has been discussed at great length and detail so many times, let me please give you a brief overview and then we can chat offline for additional details:

- Established convention for TrackXxx methods is that the thing being tracked will be sent to ingestion. However, in this case the item being tracked will not be sent. Now when the method is called, nor later. It will merely contribute to an aggregation that will be sent at some future time. So having this method aggregate would give it an unexpected behavior. We would be immediately responding to tons of customer requests of the kind "I tracked the metric value X, why do you not show it to me that way?".

- Changing the behavior of an existing API is a highly discouraged practice.

- One of the requirements is that thousands of metric measurements may be observed per second. We need APIs that allow contributing values to aggregates without object creation and lookups. This requires "handles" to metric series.

- Local dimension capping is required to protect against memory problems due to unthoughtful telemetry collection.

- Different kinds of aggregation should be supported.

- QuickPulse integration must be possible.

- Multiple dimensions must be possible.

We have evaluated multiple options. This includes your suggestion, separate, strongly typed metric classes, the present solution, where metrics are configured in one place and tracked all over the code, as well as other options. While there is no ideal solution to a complex problem, the feature team believed that this was the most optimal balance.

I would be happy to chat more. :)